gcloud-java-core/src/main/java/com/google/gcloud/BaseWriteChannel.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud;
+
+import com.google.common.base.MoreObjects;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Base implementation for a {@link WriteChannel}.
+ *
+ * @param <ServiceOptionsT> the service options used by the channel to issue RPC requests
+ * @param <EntityT> the entity this channel writes data to. Possibly with additional configuration
+ */
+public abstract class BaseWriteChannel<
+    ServiceOptionsT extends ServiceOptions<?, ?, ServiceOptionsT>,
+    EntityT extends Serializable> implements WriteChannel {
+
+  private static final int MIN_CHUNK_SIZE = 256 * 1024;
+  private static final int DEFAULT_CHUNK_SIZE = 8 * MIN_CHUNK_SIZE;
+
+  protected final ServiceOptionsT options;
```

Comment:
To prevent bad practice or mishandling in sub-classes I suggest to have all members private with protected accessors.
This is not an issue for the final members but I would favor consistency.